homeassistant/components/ambient_station/__init__.py

```diff
@@ -0,0 +1,214 @@
+
+SENSOR_TYPES = {
+    '24hourrainin': ['24 Hr Rain', 'in'],
+    'baromabsin': ['Abs Pressure', 'inHg'],
+    'baromrelin': ['Rel Pressure', 'inHg'],
+    'battout': ['Battery', ''],
+    'co2': ['co2', 'ppm'],
+    'dailyrainin': ['Daily Rain', 'in'],
+    'dewPoint': ['Dew Point', ['°F', '°C']],
+    'eventrainin': ['Event Rain', 'in'],
+    'feelsLike': ['Feels Like', ['°F', '°C']],
+    'hourlyrainin': ['Hourly Rain Rate', 'in/hr'],
+    'humidity': ['Humidity', '%'],
+    'humidityin': ['Humidity In', '%'],
+    'lastRain': ['Last Rain', ''],
+    'maxdailygust': ['Max Gust', 'mph'],
+    'monthlyrainin': ['Monthly Rain', 'in'],
+    'solarradiation': ['Solar Rad', 'W/m^2'],
+    'tempf': ['Temp', ['°F', '°C']],
+    'tempinf': ['Inside Temp', ['°F', '°C']],
+    'totalrainin': ['Lifetime Rain', 'in'],
+    'uv': ['uv', 'Index'],
+    'weeklyrainin': ['Weekly Rain', 'in'],
+    'winddir': ['Wind Dir', '°'],
+    'winddir_avg10m': ['Wind Dir Avg 10m', '°'],
+    'winddir_avg2m': ['Wind Dir Avg 2m', 'mph'],
+    'windgustdir': ['Gust Dir', '°'],
+    'windgustmph': ['Wind Gust', 'mph'],
+    'windspdmph_avg10m': ['Wind Avg 10m', 'mph'],
+    'windspdmph_avg2m': ['Wind Avg 2m', 'mph'],
+    'windspeedmph': ['Wind Speed', 'mph'],
+    'yearlyrainin': ['Yearly Rain', 'in'],
+}
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN:
+        vol.Schema({
+            vol.Required(CONF_APP_KEY):
+                cv.string,
+            vol.Required(CONF_API_KEY):
+                cv.string,
+            vol.Optional(
+                CONF_MONITORED_CONDITIONS, default=list(SENSOR_TYPES)):
+                vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+            vol.Optional(CONF_UNIT_SYSTEM):
+                vol.In([UNITS_SI, UNITS_US]),
+        })
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Set up the Ambient PWS component."""
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][DATA_CLIENT] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+
+    if conf[CONF_APP_KEY] in configured_instances(hass):
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={'source': SOURCE_IMPORT}, data=conf))
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up the Ambient PWS as config entry."""
+    from aioambient import Client
+    from aioambient.errors import WebsocketConnectionError
+
+    session = aiohttp_client.async_get_clientsession(hass)
+
+    try:
+        ambient = AmbientStation(
+            hass,
+            config_entry,
+            Client(
+                config_entry.data[CONF_API_KEY],
+                config_entry.data[CONF_APP_KEY], session),
+            config_entry.data.get(
+                CONF_MONITORED_CONDITIONS, list(SENSOR_TYPES)),
+            config_entry.data.get(CONF_UNIT_SYSTEM))
+        hass.async_create_task(ambient.ws_connect())
```

Comment:
Could this block setup phase from finishing if the initial connect would fail and we would enter the reconnect loop?

Reply:
Great point; let me think on that.

Reply:
An alternative is to schedule directly on the loop, with `hass.loop.create_task`.